pkg/sbom/spdx/marshal.go

```diff
@@ -389,11 +400,97 @@ func (m *Marshaler) spdxAnnotations(c *core.Component, timeNow string) []spdx.An
 	return annotations
 }
 
-func (m *Marshaler) spdxLicense(c *core.Component) string {
+func (m *Marshaler) spdxLicense(c *core.Component) (string, []*spdx.OtherLicense) {
+	// Only library components contain licenses
+	if c.Type != core.TypeLibrary {
+		return "", nil
+	}
 	if len(c.Licenses) == 0 {
-		return noAssertionField
+		return noAssertionField, nil
+	}
+	return m.normalizeLicenses(c.Licenses)
+}
+
+func (m *Marshaler) normalizeLicenses(licenses []string) (string, []*spdx.OtherLicense) {
+	var otherLicenses = make(map[string]*spdx.OtherLicense) // licenseID -> OtherLicense
+
+	// Save text licenses as OtherLicense
+	for i, license := range licenses {
+		if strings.HasPrefix(license, licensing.LicenseTextPrefix) {
+			otherLicense := m.newOtherLicense(strings.TrimPrefix(license, licensing.LicenseTextPrefix), true)
+			otherLicenses[otherLicense.LicenseIdentifier] = otherLicense
+			licenses[i] = otherLicense.LicenseIdentifier
+		}
+	}
+
+	license := strings.Join(lo.Map(licenses, func(license string, index int) string {
+		// e.g. GPL-3.0-with-autoconf-exception
+		license = strings.ReplaceAll(license, "-with-", " WITH ")
+		license = strings.ReplaceAll(license, "-WITH-", " WITH ")
+		return fmt.Sprintf("(%s)", license)
+	}), " AND ")
+
+	normalizedLicense, err := expression.Normalize(license, licensing.NormalizeLicense, expression.NormalizeForSPDX)
+	if err != nil {
+		// Not fail on the invalid license
+		m.logger.Warn("Unable to marshal SPDX licenses", log.String("license", license))
+		return "", nil
+	}
+
+	license, others := m.processNonSpdxLicenses(normalizedLicense)
+	allOtherLicenses := lo.Assign(otherLicenses, others)
+
+	return license, lo.Ternary(len(allOtherLicenses) > 0, lo.Values(allOtherLicenses), nil)
+}
+
+// newOtherLicense create new OtherLicense for license not included in the SPDX license list
+func (m *Marshaler) newOtherLicense(license string, text bool) *spdx.OtherLicense {
+	otherLicense := spdx.OtherLicense{
+		// ExtractedText and LicenseName are mandatory fields
+		ExtractedText: noAssertionField,
+		LicenseName:   noAssertionField,
+	}
+	if text {
+		otherLicense.ExtractedText = license
+	} else {
```

Comment:
There are cases (e.g. `license` field from python METADATA files) when we can't understand that license is name/id/text.
We tried to detect `text` of license:
https://github.com/aquasecurity/trivy/blob/f6acdf713991f8ffdbe765178fcb8a9cde433cba/pkg/licensing/normalize.go#L596-L617

So Trivy split license name/id and license text(`text://` prefix).
That is why i used both `LicenseName` and `ExtractedText` fields:
https://github.com/DmitriyLewen/trivy/blob/f851f9bb18411db838fc65e0c6c4351b04953f8f/pkg/sbom/spdx/marshal_private_test.go#L92-L112

Another question - i used `NOASSERTION` for `LicenseName` and `ExtractedText` fields (see link above), because these fields are mandatory (https://github.com/spdx/tools-golang/blob/f6e45fdb9e4e0c993105f798bee5f8aa8ea70f84/spdx/v2/v2_3/other_license.go#L12-L20).
Is this correct?

Reply:
> So Trivy split license name/id and license text(text:// prefix).
> That is why i used both LicenseName and ExtractedText fields

Makes sense - I wasn't sure how the text was captured.  Since you are capturing the text and name distinctly, your approach should work.  BTW, it's a bit tricky to find the start and end of the license text and even trickier to match it to know licenses - something we've been working on in the SPDX java libraries for about 10 years and still don't have it perfected ;).

> i used NOASSERTION for LicenseName and ExtractedText fields (see link above), because these fields are mandatory (https://github.com/spdx/tools-golang/blob/f6e45fdb9e4e0c993105f798bee5f8aa8ea70f84/spdx/v2/v2_3/other_license.go#L12-L20).
Is this correct?

For the license name, this is OK.  The spec isn't very clear on how these should be treated, so many people make up a name based on the text.  Unlike other parts of the spec, NOASSERTION is not required if you don't know - but in this case I think it would be fine to use NOASSERTION - for the name.

For the text, I would put in whatever string was actually found - even if it is the name.  The definition of the field is the license text found - so if someone puts in "This software is licensed under mylicense" - you can put that exact string in the text field even though it is not technically the text of "mylicense".  I wouldn't use "NOASSERTION" for the text field.

Reply:
got it!
Updated text field in https://github.com/aquasecurity/trivy/pull/8077/commits/659f992f9eccbfec55ea65fe7b2bdc973d892c8e